pkg/apis/provisioning/v1alpha2/provisioner_defaults.go

```diff
@@ -16,10 +16,67 @@ package v1alpha2
 
 import (
 	"context"
+	"encoding/base64"
+	"errors"
+	"io/ioutil"
+	"os"
+
+	"go.uber.org/zap"
 )
 
 // SetDefaults for the provisioner
 func (p *Provisioner) SetDefaults(ctx context.Context) {}
 
 // SetDefaults for the provisioner, cascading to all subspecs
 func (s *ProvisionerSpec) SetDefaults(ctx context.Context) {}
+
+// WithDefaults returns a copy of this Provisioner with some empty/missing
+// properties replaced by (potentially dynamic) cloud provider agnostic default values.
+// The returned copy might be complemented by dynamic default values which
+// must not be hoisted (saved) into the original Provisioner CRD as those
+// default values might change over time (e.g. rolling upgrade of CABundle, ...).
+func (p *Provisioner) WithDynamicDefaults() (_ Provisioner, err error) {
+	provisioner := *p.DeepCopy()
+	provisioner.Spec, err = provisioner.Spec.withDynamicDefaults()
+	return provisioner, err
+}
+
+// WithDefaults returns a copy of this ProvisionerSpec with some empty
+// properties replaced by default values.
+func (s *ProvisionerSpec) withDynamicDefaults() (_ ProvisionerSpec, err error) {
+	spec := *s.DeepCopy()
+	spec.Cluster, err = spec.Cluster.withDynamicDefaults()
+	return spec, err
+}
+
+// WithDefaults returns a copy of this Cluster with some empty
+// properties replaced by default values. Notably, it will try
+// to load the CABundle from the in-cluster configuraiton if it
+// is not explicitly set.
+func (c *Cluster) withDynamicDefaults() (_ Cluster, err error) {
+	cluster := *c.DeepCopy()
+	cluster.CABundle, err = cluster.getCABundle()
+	return cluster, err
+}
+
+func (c *Cluster) getCABundle() (*string, error) {
+	if c.CABundle != nil {
+		// If CABundle is explicitly provided use that one. An empty string is
+		// a valid value here if no CABundle is required at all.
+		return c.CABundle, nil
+	} else {
```

Comment:
Since the first statement uses a shortcircuiting pattern, we can remove the else and unindent all of this logic.